File: Implementierung/Backend/src/main/java/edu/kit/informatik/model/userdata/courses/Session.java
package edu.kit.informatik.model.userdata.courses;

import edu.kit.informatik.model.DataObject;
import edu.kit.informatik.model.User;
import edu.kit.informatik.model.userdata.interactions.Interaction;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.Table;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.Cascade;
import org.hibernate.annotations.CascadeType;

import java.sql.Timestamp;
import java.util.ArrayList;
import java.util.List;

/**
 * Entität einer Sitzung
 *
 * @author uekai
 * @author ugqbo
 * @version 1.0
 */
@Getter
@NoArgsConstructor
@Entity
@Table(name = "sessions")
public class Session extends DataObject {

    private String name;
    private String date;
    @ManyToOne
    private Course course;

    @OneToMany(fetch = FetchType.EAGER)
    private List<Interaction> interactions;

    @Cascade(CascadeType.SAVE_UPDATE)
    @ManyToOne
    private SeatArrangement seatArrangement;

    /**
     * Konstruktor zum Erstellen eines Objektes
     * @param user {@link User}
     * @param name Name
     * @param date Datum
     * @param course {@link Course}
     * @param seatArrangement {@link SeatArrangement}
     * @param createdAt {@link Timestamp} der Erstellung
     * @param updatedAt {@link Timestamp} der letzten Änderung
     */
    public Session(User user, String name, String date, Course course,
                   SeatArrangement seatArrangement, Timestamp createdAt, Timestamp updatedAt) {
        super(user, createdAt, updatedAt);
        this.name = name;
        this.date = date;
        this.course = course;
        this.seatArrangement = seatArrangement;
        this.interactions = new ArrayList<>();
    }

    /**
     * Setzen des Namens
     * @param name Name
     */
    public void setName(String name) {
        this.name = name;
    }

    /**
     * Setzen der {@link SeatArrangement Sitzordnmung}
     * @param seatArrangement {@link SeatArrangement Sitzordnmung}
     */
    public void setSeatArrangement(SeatArrangement seatArrangement) {
        this.seatArrangement = seatArrangement;
    }

    /**
     * Setzen der {@link Interaction Interaktionen}
     * @param interactions {@link List} von {@link Interaction Interaktionen}
     */
    public void setInteractions(List<Interaction> interactions) {
        this.interactions = interactions;
    }

    /**
     * Hinzufügen einer {@link Interaction Interaktion}
     * @param interaction {@link Interaction Interaktion}
     */
    public void addInteraction(Interaction interaction) {
        this.interactions.add(interaction);
    }

    /**
     * Entfernen einer {@link Interaction Interaktion}
     * @param interaction {@link Interaction Interaktion}
     */
    public void removeInteraction(Interaction interaction) {
        this.interactions.remove(interaction);
    }
}

File: Implementierung/frontend/paschi/src/model/userdata/courses/Session.ts
import {User} from "@/model/User";
import {Interaction} from "@/model/userdata/interactions/Interaction";
import {Course} from "@/model/userdata/courses/Course";
import {SeatArrangement} from "@/model/userdata/courses/SeatArrangement";
import {DataObject} from "@/model/DataObject";

export class Session extends DataObject {

  private readonly _user: User;
  private _name: string;
  private _date: string;
  private readonly _course: Course;
  private _interactions: Interaction[];
  private _undoInteractions: Interaction[];
  private _seatArrangement: SeatArrangement;

  constructor(id: string | undefined, localId: number, user: User, name: string, date: string, course: Course,
              seatArrangement: SeatArrangement) {
    super(id, localId);
    this._user = user;
    this._date = date;
    this._name = name;
    this._course = course;
    this._seatArrangement = seatArrangement;
    this._interactions = [];
    this._undoInteractions = [];
  }

  addInteraction(interaction: Interaction) {
    if (this.getInteraction(interaction.getId) == undefined) {
      this._interactions.push(interaction);
      this._undoInteractions = [];
    }
    this.update();
  }

  removeInteraction(interactionId: string) {
    this._interactions.forEach((element: Interaction, index: number) => {
      if (element.getId === interactionId) {
        this._interactions.splice(index, 1);
      }
    });
    this.update();
  }

  undoInteraction(): Interaction | undefined {
    const interaction = this.interactions.pop();
    if (interaction) {
      this._undoInteractions.push(interaction);
      this.update();
      return interaction;
    }

    return undefined;
  }

  redoInteraction(): Interaction | undefined {
    const interaction = this._undoInteractions.pop();
    if (interaction) {
      this.interactions.push(interaction);
      this.update();
      return interaction;
    }
    return undefined;
  }

  hasRedo(): boolean {
    return this._undoInteractions.length != 0;
  }

  hasUndo(): boolean {
    return this.interactions.length != 0;
  }

  getInteraction(interactionId: string): Interaction | undefined {
    for (const interaction of this._interactions) {
      if (interaction.getId === interactionId) {
        return interaction;
      }
    }
    return undefined;
  }

  get user(): User {
    return this._user;
  }

  get course(): Course {
    return this._course;
  }

  get name(): string {
    return this._name;
  }

  set name(value: string) {
    this._name = value;
    this.update();
  }

  get date(): string {
    return this._date;
  }

  set date(value: string) {
    this._date = value;
    this.update();
  }

  get interactions(): Interaction[] {
    return this._interactions;
  }

  set interactions(value: Interaction[]) {
    this._interactions = value;
    this.update();
  }

  get seatArrangement(): SeatArrangement {
    return this._seatArrangement;
  }

  set seatArrangement(value: SeatArrangement) {
    this._seatArrangement = value;
    this.update();
  }
}
